pkg/brokerpak/fetch.go

```diff
@@ -25,7 +40,91 @@ func fetchArchive(src, dest string) error {
 		Src:           src,
 		Dst:           dest,
 		Mode:          getter.ClientModeFile,
-		Getters:       getter.Getters,
 		Decompressors: map[string]getter.Decompressor{},
 	}).Get()
 }
+
+// fetchBrokerpak downloads a local or remote brokerpak; brokerpaks can be
+// fetched remotely using the gs:// prefix which will load them from a
+// Cloud Storage bucket with the broker's credentials.
+// Relative paths are resolved relative to the executable.
+func fetchBrokerpak(src, dest string) error {
+	execWd := filepath.Dir(os.Args[0])
+	execDir, err := filepath.Abs(execWd)
+	if err != nil {
+		return fmt.Errorf("couldn't turn dir %q into abs path: %v", execWd, err)
+	}
+
+	getters := map[string]getter.Getter{}
+	for k, g := range getter.Getters {
+		getters[k] = g
+	}
+	getters["gs"] = &gsGetter{}
+
+	return (&getter.Client{
+		Src: src,
+		Dst: dest,
+		Pwd: execDir,
+
+		Mode:          getter.ClientModeFile,
+		Getters:       getters,
+		Decompressors: map[string]getter.Decompressor{},
+	}).Get()
+}
+
+// gsGetter is a go-getter that works on Cloud Storage using the broker's
+// service account. It's incomplete in that it doesn't support directories.
+type gsGetter struct{}
+
+// ClientMode is unsupported for gsGetter.
+func (g *gsGetter) ClientMode(u *url.URL) (getter.ClientMode, error) {
+	return getter.ClientModeInvalid, errors.New("mode is not supported for this client")
+}
+
+// Get clones a remote destination
+func (g *gsGetter) Get(dst string, u *url.URL) error {
+	return errors.New("getting directories is not supported for this client")
+}
+
+// GetFile downloads the give URL into the given path. The URL must
+// reference a single file. If possible, the Getter should check if
+// the remote end contains the same file and no-op this operation.
+func (g *gsGetter) GetFile(dst string, u *url.URL) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+
+	obj, err := g.objectAt(u)
+	if err != nil {
+		return err
+	}
+
+	src := stream.FromReadCloserError(obj.NewReader(ctx))
```

Comment:
I think the error on `obj.NewReader(ctx)` should be checked and raised, instead of silenced. Because of how the storage package is written, no "invalid bucket" or "invalid object" errors are thrown until the read call. So the src could be invalid (or empty, most likely) and `GetFile` might return no error.